Dotnet exporters update 2 (#4016)
open-telemetry/opentelemetry.io: content/en/docs/languages/net/exporters.md

{{% docs/languages/exporters/intro dotnet %}}

### Dependencies {#otlp-dependencies}

If you want to send telemetry data to an OTLP endpoint (like the
[OpenTelemetry Collector](#collector-setup), [Jaeger](#jaeger) or
[Prometheus](#prometheus)), you can choose between two different protocols to
transport your data:

- HTTP/protobuf
- gRPC

> Start by installing the respective exporter package as a dependency for your

Reviewer: nit: We can simply say "Install the OpenTelemetry.Exporter.OpenTelemetryProtocol package".. The "respective" package gives the impression that there are diff. exporter packages for HTTP/protobuf vs gRPC!

> Start by installing the respective exporter package as a dependency for your
> project:

Author: ```suggestion
Start by installing the [`OpenTelemetry.Exporter.OpenTelemetryProtocol`](https://www.nuget.org/packages/OpenTelemetry.Exporter.OpenTelemetryProtocol/) package as a dependency for your
project:
```

```sh
dotnet add package OpenTelemetry.Exporter.OpenTelemetryProtocol
dotnet add package OpenTelemetry.Extensions.Hosting
```

### Usage

> If you're using ASP.NET Core, configure the exporters in your ASP.NET Core

Reviewer: since we are splitting this based on ASP.NET Core vs others, perhaps it is better to move the 
`dotnet add package OpenTelemetry.Extensions.Hosting` as well to here.

Author: Not sure what this means, since my knowledge about .NET is limited, can you provide me with some suggestions what needs to be changed?

Author: (or can this be fixed in a follow up PR?)

Reviewer: The usage section here starts with "if you are using ASP.NET Core".. And then later there is an "else" branch. It basically means, the doc is effectively split between asp.net core and non-asp.netcore. But the contents are following this principle fully. 
Specific example in this: `OpenTelemetry.Extensions.Hosting` package is only for Asp.Net Core apps. (technically any apps that supports Host model, that include worker apps too). So my suggestion was to move that package mention into the "If asp.net core" part.

Reviewer 2: I agree with @cijothomas here. The hosting extensions package is almost exclusively relevant for ASP.NET Core apps.

Author: I understand, I can rewrite this

Author: please take another look

  {
    b.AddOtlpExporter()
    // The rest of your setup code goes here too
  })
  .WithMetrics(metrics => {
    metrics.AddOtlpExporter
    // The rest of your setup code goes here
  });

builder.Logging

>   .AddOpenTelemetry(options => {
>         options.AddOtlpExporter()

Reviewer 3: ```suggestion
  .AddOpenTelemetry(logging => {
        logging.AddOtlpExporter()
```

  {
    b.AddOtlpExporter()
    // The rest of your setup code goes here too
  })

>   .WithMetrics(b => {
>     b.AddOtlpExporter
>     // The rest of your setup code goes here
>   });

Reviewer 3: ```suggestion
  .WithMetrics(metrics => {
    metrics.AddOtlpExporter
    // The rest of your setup code goes here
  });
```

builder.Logging
  .AddOpenTelemetry(options => {
        options.AddOtlpExporter()
  });
```

This will, by default, send telemetry using gRPC to <http://localhost:4317>, to
customize this to use HTTP and the protobuf format, you can add options like
this:

```csharp
var builder = WebApplication.CreateBuilder(args);

builder.Services.AddOpenTelemetry()
  .WithTracing(b => {
    b
    .AddOtlpExporter(opt =>
    {
        opt.Endpoint = new Uri("your-endpoint-here/v1/traces");
        opt.Protocol = OtlpExportProtocol.HttpProtobuf;
    })
    // The rest of your setup code goes here too
  })

>   .WithMetrics(b => {
>     b
>     .AddOtlpExporter(opt =>
>     {
>         opt.Endpoint = new Uri("your-endpoint-here/v1/metrics");
>         opt.Protocol = OtlpExportProtocol.HttpProtobuf;
>     })

Reviewer 3: ```suggestion
  .WithMetrics(metrics => metrics
    .AddOtlpExporter(options =>
    {
        options.Endpoint = new Uri("your-endpoint-here/v1/metrics");
        options.Protocol = OtlpExportProtocol.HttpProtobuf;
    })
```

    // The rest of your setup code goes here too
  });

builder.Logging

>   .AddOpenTelemetry(options => {
>         options.AddOtlpExporter(opt =>
>         {
>             opt.Endpoint = new Uri("your-endpoint-here/v1/logs");
>             opt.Protocol = OtlpExportProtocol.HttpProtobuf;
>         })
>   });

Reviewer 3: ```suggestion
  .AddOpenTelemetry(logging => {
        logging.AddOtlpExporter(options =>
        {
            options.Endpoint = new Uri("your-endpoint-here/v1/logs");
            options.Protocol = OtlpExportProtocol.HttpProtobuf;
        })
  });
```

    // The rest of your setup code goes here
  });

builder.Logging
  .AddOpenTelemetry(logging => {
        logging.AddOtlpExporter()
  });
```

This will, by default, send telemetry using gRPC to <http://localhost:4317>, to
customize this to use HTTP and the protobuf format, you can add options like
this:

```csharp
var builder = WebApplication.CreateBuilder(args);

builder.Services.AddOpenTelemetry()
  .WithTracing(b => {
    b
    .AddOtlpExporter(opt =>
    {
        opt.Endpoint = new Uri("your-endpoint-here/v1/traces");
        opt.Protocol = OtlpExportProtocol.HttpProtobuf;
    })
    // The rest of your setup code goes here too
  })
  .WithMetrics(metrics => metrics
    .AddOtlpExporter(options =>
    {
        options.Endpoint = new Uri("your-endpoint-here/v1/metrics");
        options.Protocol = OtlpExportProtocol.HttpProtobuf;
    })
    // The rest of your setup code goes here too
  });

builder.Logging
  .AddOpenTelemetry(logging => {
        logging.AddOtlpExporter(options =>
        {
            options.Endpoint = new Uri("your-endpoint-here/v1/logs");
            options.Protocol = OtlpExportProtocol.HttpProtobuf;
        })
  });
```

#### Non-ASP.NET Core

> Configure the exporter when creating a tracer provider:

Reviewer: ```suggestion
Configure the exporter when creating a `TracerProvider`, `MeterProvider` or `LoggerFactory`:
```

    // Other setup code, like setting a resource goes here too

    .Build();

using var meterProvider = Sdk.CreateMeterProviderBuilder()

>     .AddOtlpExporter(opt =>
>     {
>         opt.Endpoint = new Uri("your-endpoint-here/v1/metrics");
>         opt.Protocol = OtlpExportProtocol.HttpProtobuf;
>     })

Reviewer 3: ```suggestion
    .AddOtlpExporter(options =>
    {
        options.Endpoint = new Uri("your-endpoint-here/v1/metrics");
        options.Protocol = OtlpExportProtocol.HttpProtobuf;
    })
```

    // Other setup code, like setting a resource goes here too

    .Build();

using var loggerFactory = LoggerFactory.Create(builder =>
{

>     builder.AddOpenTelemetry(options =>
>     {
>         options.AddOtlpExporter(opt =>
>         {
>             opt.Endpoint = new Uri("your-endpoint-here/v1/logs");
>             opt.Protocol = OtlpExportProtocol.HttpProtobuf;
>         })

Reviewer 3: ```suggestion
    builder.AddOpenTelemetry(logging =>
    {
        logging.AddOtlpExporter(options =>
        {
            options.Endpoint = new Uri("your-endpoint-here/v1/logs");
            options.Protocol = OtlpExportProtocol.HttpProtobuf;
        })
```

    // Other setup code, like setting a resource goes here too

    .Build();

> using var meterProvider = Sdk.CreateMeterProviderBuilder()

Reviewer: It'd be good to get a follow up PR to remove the pattern of "using var meterProvider", and prefer an explicit dispose().. We have a lot of people copying this to helper functions, and accidently causing it to get disposed() when helper method exits!

https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/docs/trace/getting-started-console/Program.cs#L27-L29

Separate PR to not further delay this PR

    .AddOtlpExporter(options =>
    {
        options.Endpoint = new Uri("your-endpoint-here/v1/metrics");
        options.Protocol = OtlpExportProtocol.HttpProtobuf;
    })

    // Other setup code, like setting a resource goes here too

    .Build();

using var loggerFactory = LoggerFactory.Create(builder =>
{
    builder.AddOpenTelemetry(logging =>
    {
        logging.AddOtlpExporter(options =>
        {
            options.Endpoint = new Uri("your-endpoint-here/v1/logs");
            options.Protocol = OtlpExportProtocol.HttpProtobuf;
        })
    });
});
```

Use environment variables to set values like headers and an endpoint URL for
production.

> #### Note for .NET Core 3.1 and below and gRPC

Reviewer: Given .NET Core 3.1 is out-of-support, and not supported by OTel .NET, lets remove it.

Author: removed

The console exporter is useful for development and debugging tasks, and is the
simplest to set up. Start by installing the
[`OpenTelemetry.Exporter.Console`](https://www.nuget.org/packages/OpenTelemetry.Exporter.Console/)
package as a dependency for your project:

```sh
dotnet add package OpenTelemetry.Exporter.Console
```

If you're using ASP.NET Core install the
[`OpenTelemetry.Extensions.Hosting`](https://www.nuget.org/packages/OpenTelemetry.Extensions.Hosting)
package as well:

```sh
dotnet add package OpenTelemetry.Extensions.Hosting
```

### Usage {#console-usage}

#### ASP.NET Core {#console-usage-asp-net-core}

Configure the exporter in your ASP.NET Core services:

```csharp
var builder = WebApplication.CreateBuilder(args);

builder.Services.AddOpenTelemetry()
    .WithTracing(tracing =>
    {
        tracing.AddConsoleExporter()
        // The rest of your setup code goes here too
    })
    .WithMetrics(metrics =>
        metrics.AddConsoleExporter()
        // The rest of your setup code goes here too
    });

builder.Logging.AddOpenTelemetry(logging => {
        logging.AddConsoleExporter()
  });
```

#### Non-ASP.NET Core {#console-usage-non-asp-net-core}

> Configure the exporter when creating a tracer provider:

Reviewer: ```suggestion
Configure the exporter when creating a `TracerProvider`, `MeterProvider` or `LoggerFactory`:
```

```csharp
using var tracerProvider = Sdk.CreateTracerProviderBuilder()
    .AddConsoleExporter()

    // Other setup code, like setting a resource goes here too

    .Build();

using var meterProvider = Sdk.CreateMeterProviderBuilder()
    .AddConsoleExporter()

    // Other setup code, like setting a resource goes here too

    .Build();

using var loggerFactory = LoggerFactory.Create(builder =>
{
    builder.AddOpenTelemetry(logging =>
    {
        logging.AddConsoleExporter()
    });
});
```

{{% docs/languages/exporters/jaeger %}}

{{% docs/languages/exporters/prometheus-setup %}}

### Dependencies {#prometheus-dependencies}

> Install the

Reviewer: move this also under asp.net core section.

Author: I want to keep the existing pattern of "Dependencies" and "Usage", it's a little bit odd for .NET now, but this is for consistency with other languages, an alternative would be to split it out in core vs non-core first and then go into dependencies and usage, but that might be a separate PR

Author: going to merge this PR now we can have incremental changes to improve

var builder = WebApplication.CreateBuilder(args);

builder.Services.AddOpenTelemetry()
    .WithTracing(tracing =>
    {
        tracing.AddConsoleExporter()
        // The rest of your setup code goes here too
    })
    .WithMetrics(metrics =>
        metrics.AddConsoleExporter()
        // The rest of your setup code goes here too
    });

> builder.Logging
>   .AddOpenTelemetry(logging => {

Reviewer 3: ```suggestion
builder.Logging.AddOpenTelemetry(logging => {
```

builder.Services.AddOpenTelemetry()

>     .WithTracing(b =>

Reviewer 3: ```suggestion
    .WithTracing(tracing =>
```

    {

>         b.AddConsoleExporter()

Reviewer 3: ```suggestion
        tracing.AddConsoleExporter()
```

        // The rest of your setup code goes here too
    })

>     .WithMetrics(b =>
>         b.AddConsoleExporter()

Reviewer 3: ```suggestion
    .WithMetrics(metrics =>
        metrics.AddConsoleExporter()
```

        // The rest of your setup code goes here too
    });

builder.Logging

>   .AddOpenTelemetry(options => {
>         options.AddConsoleExporter()

Reviewer 3: ```suggestion
  .AddOpenTelemetry(logging => {
        logging.AddConsoleExporter()
```

Further details on configuring the Prometheus exporter can be found
[here](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Exporter.Prometheus.AspNetCore/README.md).

{{% docs/languages/exporters/zipkin-setup %}}

### Dependencies {#zipkin-dependencies}

To send your trace data to [Zipkin](https://zipkin.io/), install the
[exporter package](https://www.nuget.org/packages/OpenTelemetry.Exporter.Zipkin)
as a dependency for your application:

```shell
dotnet add package OpenTelemetry.Exporter.Zipkin
dotnet add package OpenTelemetry.Extensions.Hosting
```

If you're using ASP.NET Core, configure the exporter in your ASP.NET Core
services:

```csharp
var builder = WebApplication.CreateBuilder(args);

builder.Services.AddOpenTelemetry()

>     .WithTracing(b =>
>     {
>         b.AddZipkinExporter(o =>
>         {
>             o.Endpoint = new Uri("your-zipkin-uri-here");
>         })

Reviewer 3: ```suggestion
    .WithTracing(tracing =>
    {
        tracing.AddZipkinExporter(options =>
        {
            options.Endpoint = new Uri("your-zipkin-uri-here");
        })
```

        // The rest of your setup code goes here too
    });
```

Otherwise, configure the exporter when creating a tracer provider:

```csharp
using var tracerProvider = Sdk.CreateTracerProviderBuilder()

>     .AddZipkinExporter(o =>
>     {
>         o.Endpoint = new Uri("your-zipkin-uri-here");

Reviewer 3: ```suggestion
    .AddZipkinExporter(options =>
    {
        options.Endpoint = new Uri("your-zipkin-uri-here");
```

await app.RunAsync();
```

#### Non-ASP.NET Core {#prometheus-non-asp-net-core-usage}

For applications not using ASP.NET Core, you can use the `HttpListener` version

> which is available in a

Reviewer: could we include a warning that this component has no intention of getting PROD ready:
https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Exporter.Prometheus.HttpListener#prometheus-exporter-httplistener-for-opentelemetry-net
